File: src/code_intelligence/classifiers/layer_classifier.py
"""Deterministic layer classifier for code intelligence graph nodes."""

from __future__ import annotations

import re
from typing import Any, Sequence

from code_intelligence.models.graph import GraphNode, NodeKind, SourceLocation

_FRONTEND_NODE_KINDS = {NodeKind.COMPONENT, NodeKind.HOOK}
_BACKEND_NODE_KINDS = {NodeKind.GUARD, NodeKind.MIDDLEWARE, NodeKind.ENDPOINT, NodeKind.REPOSITORY, NodeKind.DATABASE_CONNECTION, NodeKind.QUERY}
_INFRA_NODE_KINDS = {NodeKind.INFRA_RESOURCE, NodeKind.AZURE_RESOURCE, NodeKind.AZURE_FUNCTION}
_INFRA_LANGUAGES = {"terraform", "bicep", "dockerfile"}
_SHARED_NODE_KINDS = {NodeKind.CONFIG_FILE, NodeKind.CONFIG_KEY, NodeKind.CONFIG_DEFINITION}

_FRONTEND_PATH_RE = re.compile(
    r"(?:^|/)(?:src/)?(?:components|pages|views|app/ui|public)/",
)
_BACKEND_PATH_RE = re.compile(
    r"(?:^|/)(?:src/)?(?:server|api|controllers|services|routes|handlers)/",
)
_FRONTEND_EXT_RE = re.compile(r"\.(?:tsx|jsx)$")

_FRONTEND_FRAMEWORKS = {"react", "vue", "angular", "svelte", "nextjs"}
_BACKEND_FRAMEWORKS = {"express", "nestjs", "flask", "django", "fastapi", "spring"}


class LayerClassifier:
    """Assigns a deterministic 'layer' property to every graph node.
    Rules are evaluated in order; first match wins.
    """

    def classify(self, nodes: Sequence[GraphNode]) -> None:
        for node in nodes:
            node.properties["layer"] = self._classify_one(node)

    def classify_store(self, store: Any) -> None:
        """Classify nodes directly in a GraphStore, updating networkx data in-place."""
        for node_id, data in store.graph.nodes(data=True):
            if "kind" not in data:
                continue
            # Build a lightweight GraphNode for classification
            loc = data.get("location")
            if isinstance(loc, dict):
                loc = SourceLocation(**loc)
            node = GraphNode(
                id=data.get("id", node_id),
                kind=NodeKind(data["kind"]),
                label=data.get("label", ""),
                location=loc,
                properties=data.get("properties", {}),
            )
            layer = self._classify_one(node)
            # Update the networkx data dict directly
            props = data.setdefault("properties", {})
            props["layer"] = layer

    def _classify_one(self, node: GraphNode) -> str:
        if node.kind in _FRONTEND_NODE_KINDS:
            return "frontend"
        if node.kind in _BACKEND_NODE_KINDS:
            return "backend"
        if node.kind in _INFRA_NODE_KINDS:
            return "infra"
        lang = node.properties.get("language", "")
        if lang in _INFRA_LANGUAGES:
            return "infra"
        file_path = ""
        if node.location:
            file_path = node.location.file_path
        if _FRONTEND_EXT_RE.search(file_path):
            return "frontend"
        if _FRONTEND_PATH_RE.search(file_path):
            return "frontend"
        if _BACKEND_PATH_RE.search(file_path):
            return "backend"
        fw = node.properties.get("framework", "")
        if fw in _FRONTEND_FRAMEWORKS:
            return "frontend"
        if fw in _BACKEND_FRAMEWORKS:
            return "backend"
        if node.kind in _SHARED_NODE_KINDS:
            return "shared"
        return "unknown"

File: src/main/java/io/github/randomcodespace/iq/analyzer/LayerClassifier.java
package io.github.randomcodespace.iq.analyzer;

import io.github.randomcodespace.iq.model.CodeNode;
import io.github.randomcodespace.iq.model.NodeKind;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Set;
import java.util.regex.Pattern;

/**
 * Deterministic layer classifier for graph nodes.
 * Assigns a {@code layer} property to every node: frontend, backend,
 * infra, shared, or unknown.
 * <p>
 * Rules are evaluated in priority order; first match wins.
 */
@Service
public class LayerClassifier {

    private static final Set<NodeKind> FRONTEND_NODE_KINDS = Set.of(
            NodeKind.COMPONENT, NodeKind.HOOK
    );

    private static final Set<NodeKind> BACKEND_NODE_KINDS = Set.of(
            NodeKind.GUARD, NodeKind.MIDDLEWARE, NodeKind.ENDPOINT,
            NodeKind.REPOSITORY, NodeKind.DATABASE_CONNECTION, NodeKind.QUERY,
            NodeKind.ENTITY, NodeKind.MIGRATION, NodeKind.SERVICE,
            NodeKind.TOPIC, NodeKind.QUEUE, NodeKind.EVENT,
            NodeKind.MESSAGE_QUEUE, NodeKind.RMI_INTERFACE,
            NodeKind.WEBSOCKET_ENDPOINT
    );

    private static final Set<NodeKind> INFRA_NODE_KINDS = Set.of(
            NodeKind.INFRA_RESOURCE, NodeKind.AZURE_RESOURCE, NodeKind.AZURE_FUNCTION
    );

    private static final Set<String> INFRA_LANGUAGES = Set.of(
            "terraform", "bicep", "dockerfile"
    );

    private static final Set<NodeKind> SHARED_NODE_KINDS = Set.of(
            NodeKind.CONFIG_FILE, NodeKind.CONFIG_KEY, NodeKind.CONFIG_DEFINITION,
            NodeKind.PROTOCOL_MESSAGE
    );

    private static final Pattern FRONTEND_PATH_RE = Pattern.compile(
            "(?:^|/)(?:src/)?(?:components|pages|views|app/ui|public)/"
    );

    private static final Pattern BACKEND_PATH_RE = Pattern.compile(
            "(?:^|/)(?:src/)?(?:server|api|controllers|services|routes|handlers)/"
    );

    private static final Pattern FRONTEND_EXT_RE = Pattern.compile(
            "\\.(?:tsx|jsx)$"
    );

    private static final Set<String> FRONTEND_FRAMEWORKS = Set.of(
            "react", "vue", "angular", "svelte", "nextjs"
    );

    private static final Set<String> BACKEND_FRAMEWORKS = Set.of(
            "express", "nestjs", "flask", "django", "fastapi", "spring"
    );

    /**
     * Classify all nodes in the list, setting the {@code layer} property on each.
     */
    public void classify(List<CodeNode> nodes) {
        for (CodeNode node : nodes) {
            node.setLayer(classifyOne(node));
        }
    }

    /**
     * Classify a single node.
     */
    String classifyOne(CodeNode node) {
        // 1. Node kind rules
        if (FRONTEND_NODE_KINDS.contains(node.getKind())) return "frontend";
        if (BACKEND_NODE_KINDS.contains(node.getKind())) return "backend";
        if (INFRA_NODE_KINDS.contains(node.getKind())) return "infra";

        // 2. Language rules
        Object lang = node.getProperties().get("language");
        if (lang instanceof String langStr && INFRA_LANGUAGES.contains(langStr)) {
            return "infra";
        }

        // 3. File path rules
        String filePath = node.getFilePath() != null ? node.getFilePath() : "";
        if (FRONTEND_EXT_RE.matcher(filePath).find()) return "frontend";
        if (FRONTEND_PATH_RE.matcher(filePath).find()) return "frontend";
        if (BACKEND_PATH_RE.matcher(filePath).find()) return "backend";

        // 4. Framework rules
        Object fw = node.getProperties().get("framework");
        if (fw instanceof String fwStr) {
            if (FRONTEND_FRAMEWORKS.contains(fwStr)) return "frontend";
            if (BACKEND_FRAMEWORKS.contains(fwStr)) return "backend";
        }

        // 5. Shared node kinds
        if (SHARED_NODE_KINDS.contains(node.getKind())) return "shared";

        return "unknown";
    }
}
